vllm/worker/worker_base.py

```
@@ -60,28 +123,26 @@ def init_device(self) -> None:
         """
         raise NotImplementedError
 
-    @abstractmethod
-    def determine_num_available_blocks(self) -> Tuple[int, int]:
-        """Determine the number of available blocks for the GPU KV cache and
-        swappable CPU KV cache.
-
-        The implementation may run profiling or other heuristics to determine
-        the size of caches.
-
-        Returns a Tuple[num_gpu_blocks, num_cpu_blocks], where num_gpu_blocks
-        are blocks that are "active" on the device and can be appended to.
-        num_cpu_blocks refers to "swapped" blocks in CPU memory and cannot be
-        appended to.
-        """
-        raise NotImplementedError
-
-    @abstractmethod
     def initialize_cache(self, num_gpu_blocks: int,
                          num_cpu_blocks: int) -> None:
         """Initialize the KV cache with the given size in blocks.
         """
         raise NotImplementedError
 
+    def get_model(self) -> nn.Module:
+        raise NotImplementedError
+
+    def load_model(self) -> None:
+        """Load model onto target device."""
+        raise NotImplementedError
+
+    def execute_model(
+        self,
+        execute_model_req: Optional[ExecuteModelRequest] = None
+    ) -> Optional[List[SamplerOutput]]:
+        raise NotImplementedError
+
+    @avoid_check
```

I feel this is too complicated, we need to decorate `@avoid_check` for every method that has a common implementation in the base class.

---

We can check `__isabstractmethod__` attribute to only select the methods that are decorated by `@abstractmethod`.

---

We probably need to go back to the original point of the discussion which is making development easier, right? Thanks @youkaichao for the reminder, I initially inherited v0 workerbase (`vllm.worker.worker_base.WorkerBase`). However, this base has some `@abstractmethod` that are not needed in v1 right now, such as `add_lora`,`remove_lora`. 

It's easy if you just remove `@abstractmethod`, but it's probably not development-friendly because developers always want a good reminder for what methods aren't implemented right now. Therefore, with the following user experience, there are some unimplemented methods in the v1 GPU.

`WARNING 02-10 12:57:48 worker_base.py:67] Methods add_lora,determine_num_available_blocks,get_cache_block_size_bytes,list_loras,pin_lora,remove_lora not implemented in <vllm.v1.worker.gpu_worker.Worker object at 0x7a025b3297b0>`

As a developer, you can choose to implement it or not, and you don't need to go through trial and error to find a way to achieve it. Right? So the final experience is that we only keep the key abstract methods, and tell the developers the rest through reminders

BTW, @DarkLight1337 , we cannot use `__isabstractmethod__` since we removed most `@abstractmethod`. And there is only one `start_worker_execution_loop` needing @avoid_check, so this shouldn't be a huge amount of work? @youkaichao

---

How about keeping abstractmethod but effectively disabling them by removing `ABC` base class?

---

> How about keeping abstractmethod but effectively disabling them by removing `ABC` base class?

@DarkLight1337 this makes life easier and won't throw errors if we don't implement `@abstractmethod` in base class. **But there was no reminder of this.** If this is an approach you agree with, isn't @youkaichao 's idea a more straightforward approach that doesn't even require an `@abstractmethod`?

---

I think currently `check_implementation` replaces the functionality of `ABC` (which throws errors instead of warnings if there are unimplemented methods). Basically, we use `abstractmethod` (soft check at initialization) + `raise NotImplementedError` (hard check at call site) to remind ourselves which methods are still missing.

---

> I think currently `check_implementation` replaces the functionality of `ABC` (which throws errors instead of warnings if there are unimplemented methods). Basically, we use `abstractmethod` (soft check at initialization) + `raise NotImplementedError` (hard check at call site) to remind ourselves which methods are still missing.

I see. This makes sense to me.

---

Hey guys, I really love the open source community's discussion atmosphere, but I also hope this PR doesn't drop halfway. In this case, I don't know if you members have any mechanisms? Like voting or something? No matter what kind of change it is, I might need a clear and uniform suggestion. Thank you! @youkaichao @DarkLight1337